### ISSUE TITLE
Changes to ButinaSplitter

### ISSUE THREAD

**Issue by peastman** [deepchem/deepchem#2198]
cc @rbharath @proteneer 

This is a followup to https://github.com/deepchem/deepchem/issues/2112#issuecomment-700942341 and the discussion following it.  ButinaSplitter is currently not very useful, but could be made more useful with a few changes.

As discussed there I modified it to use the same code for building datasets as ScaffoldSplitter.  I then tested it with the following code.

```python
splitters = ['random', 'scaffold', 'butina']
metric = dc.metrics.Metric(dc.metrics.roc_auc_score)
for splitter in splitters:
    tasks, datasets, transformers = dc.molnet.load_tox21(featurizer='ECFP', split=splitter)
    train_dataset, valid_dataset, test_dataset = datasets
    model = dc.models.MultitaskClassifier(n_tasks=len(tasks), n_features=1024, layer_sizes=[1000])
    model.fit(train_dataset, nb_epoch=10)
    print('splitter:', splitter)
    print('training set score:', model.evaluate(train_dataset, [metric], transformers))
    print('test set score:', model.evaluate(test_dataset, [metric], transformers))
    print()
```

RandomSplitter gives a test set score of 0.78, and ScaffoldSplitter gives 0.68, which shows how random splitting doesn't produce a truly independent test set.  Scaffold splitting does much better.  ButinaSplitter gives a test set score of 0.78, exactly the same as RandomSplitter!

ButinaSplitter has an adjustable cutoff.  The above was with the default value of 0.18.  So I repeated the test with different values of the cutoff.  Here are the test set scores.

0.18: 0.78
0.3: 0.74
0.4: 0.73
0.5: 0.66
0.6: 0.61
0.7: 0.65
0.8: 0.61

At least on this dataset, 0.18 is much too small.  It works no better than random splitting.  It gets better as you increase the cutoff, and by 0.6 it's significantly better than ScaffoldSplitter.  Further increases make little difference.

I looked up the original paper on the clustering method (http://www.l4patterns.com/uploads/dbclus-paper.pdf), and they used a cutoff of 0.8.  How was the value 0.18 selected?  Would it be reasonable to increase the default to 0.6?  Is this value likely to be universal, or will it vary a lot between datasets?

**Comment by proteneer** [comment]
I think we'd need to test on test on different datasets to see the effect of the cutoff - I think I tested it on a proprietary dataset that was super diverse. I think increasing the default to 0.6 makes sense though - though I'd caution the user still to do some sanity checks against the Random Splitter as you have down.

**Comment by rbharath** [comment]
+1 for increasing the default to 0.6. Perhaps worth trying on a couple other MoleculeNet datasets to see if 0.6 behaves reasonably and adding some documentation directing users to do sanity checks as needed

**Comment by peastman** [comment]
I tested on a couple of other datasets.  Here are the test set scores:

BACE

random: 0.92
scaffold: 0.80
butina 0.4: 0.69
butina 0.5: 0.71
butina 0.6: 0.63
butina 0.7: 0.59
butina 0.8: 0.63

Clintox

random: 0.72
scaffold: 0.64
butina 0.4: 0.65
butina 0.5: 0.72
butina 0.6: 0.57
butina 0.7: 0.59
butina 0.8: 0.68

It looks like 0.6 is a pretty good default value.  The minimum is somewhere in that range for all three datasets, and in all cases it's better than scaffold splitting.

These are all pretty small datasets.  I'll see if I can get it to work on a larger one without crashing my computer.

**Comment by peastman** [comment]
Nope.  Tried on HIV again, which has more than five times as many samples as tox21.  It ran for about half an hour, and then the computer crashed.  Conclusion: butina is a good method for small datasets, but not for large ones.  We should look into other alternatives to see if there's one that works in a similar way to butina (cluster by fingerprints) but scales better.
